src/mlpack/methods/ann/layer/multihead_attention_impl.hpp

```diff
@@ -0,0 +1,288 @@
+MultiheadAttention(const size_t tLen,
+    const size_t sLen,
+    const size_t embedDim,
+    const size_t numHeads,
+    const ElemType dropoutRate,
+    const bool deterministic) :
+    tLen(tLen),
+    sLen(sLen),
+    dropoutRate(dropoutRate),
+    deterministic(deterministic)
+{
+  Log::Assert(embedDim % numHeads == 0,
+      "'embedDim' must be divisible by 'numHeads'");
+  this->embedDim = embedDim;
+  this->numHeads = numHeads;
+  this->headDim = embedDim / numHeads;
+
+  weights.set_size(4 * (embedDim + 1) * embedDim, 1);
+  dropout.Ratio(dropoutRate);
+  dropout.Deterministic() = deterministic;
+}
+
+template <typename InputDataType, typename OutputDataType,
+          typename RegularizerType>
+void MultiheadAttention<InputDataType, OutputDataType, RegularizerType>::
+ResetParameters()
+{
+  queryWt = arma::mat(weights.memptr(), embedDim, embedDim, false, false);
+  keyWt = arma::mat(weights.memptr() + embedDim * embedDim,
+      embedDim, embedDim, false, false);
+  valueWt = arma::mat(weights.memptr() + 2 * embedDim * embedDim,
+      embedDim, embedDim, false, false);
+  outWt = arma::mat(weights.memptr() + 3 * embedDim * embedDim,
+      embedDim, embedDim, false, false);
+
+  queryBias = arma::mat(weights.memptr()
+      + 4 * embedDim * embedDim, embedDim, 1, false, false).zeros();
+  keyBias = arma::mat(weights.memptr()
+      + (4 * embedDim + 1) * embedDim, embedDim, 1, false, false).zeros();
+  valueBias = arma::mat(weights.memptr()
+      + (4 * embedDim + 2) * embedDim, embedDim, 1, false, false).zeros();
+  outBias = arma::mat(weights.memptr()
+      + (4 * embedDim + 3) * embedDim, embedDim, 1, false, false).zeros();
+  XavierInitialization().Initialize(queryWt, embedDim, embedDim);
+  XavierInitialization().Initialize(keyWt, embedDim, embedDim);
+  XavierInitialization().Initialize(valueWt, embedDim, embedDim);
+  XavierInitialization().Initialize(outWt, embedDim, embedDim);
+}
+
+template <typename InputDataType, typename OutputDataType,
+          typename RegularizerType>
+template <typename eT>
+void MultiheadAttention<InputDataType, OutputDataType, RegularizerType>::
+Forward(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  Log::Assert(input.n_rows == embedDim * (tLen + 2 * sLen));
+  const size_t bsz = input.n_cols;
+  output.set_size(embedDim * tLen, bsz);
+
+  arma::cube query(const_cast<arma::Mat<eT>&>(input).memptr(),
+      embedDim, tLen, bsz, false, false);
+  arma::cube key(const_cast<arma::Mat<eT>&>(input).memptr()
+      + query.n_elem, embedDim, sLen, bsz, false, false);
+  arma::cube value(const_cast<arma::Mat<eT>&>(input).memptr()
+      + query.n_elem + key.n_elem, embedDim, sLen, bsz, false, false);
+
+  for (size_t i = 0; i < bsz; ++i)
+  {
+    query.slice(i) = queryWt * query.slice(i) + arma::repmat(queryBias,1, tLen);
+    key.slice(i) = keyWt * key.slice(i) + arma::repmat(keyBias, 1, sLen);
+    value.slice(i) = valueWt * key.slice(i) + arma::repmat(valueBias, 1, sLen);
```

Comment:
I think it's undefined behavior. These matrices are parts of `input` which has the `const` modifier. Therefore you can't change them.

Reply:
Ohh... right. I think I should use copy while defining those matrices.
```cpp
arma::cube query(const_cast<arma::Mat<eT>&>(input).memptr(), embedDim, tLen, bsz, true);
arma::cube key(const_cast<arma::Mat<eT>&>(input).memptr() + query.n_elem, embedDim, sLen, bsz, true);
arma::cube value(const_cast<arma::Mat<eT>&>(input).memptr() + query.n_elem + key.n_elem, embedDim, sLen, bsz, true);
```